File: javascript/src/boundary/AppPaciente.js
import { Documento } from '../dominio/Documento.js';
import { TipoDocumento } from '../dominio/TipoDocumento.js';

/**
 * Fronteira do sistema (nao e entidade de dominio).
 * Repare que ela depende SO de upload: nao tem acesso a download nem a triagem.
 */
export class AppPaciente {
  #auth;
  #agenda;
  #upload;
  #sessao = null;

  constructor(auth, agenda, upload) {
    this.#auth = auth;
    this.#agenda = agenda;
    this.#upload = upload;
  }

  entrar(email, senha) {
    this.#sessao = this.#auth.login(email, senha);
    console.log(`[app-paciente] login ok, token valido ate ${this.#sessao.expiraEm.toLocaleTimeString('pt-BR')}`);
  }

  agendarConsulta(paciente, medico, dataHora, modalidade) {
    this.#exigirSessao();
    return this.#agenda.agendar(paciente, medico, dataHora, modalidade);
  }

  enviarExamePrevio(consulta, uri, autor) {
    this.#exigirSessao();
    const documento = new Documento(TipoDocumento.EXAME_PREVIO, uri, autor);
    consulta.anexar(documento);
    return this.#upload.enviar(documento);
  }

  #exigirSessao() {
    if (!this.#sessao || !this.#sessao.estaValida()) {
      throw new Error('Faca login antes.');
    }
  }
}

File: java/src/br/unifor/facilitador/boundary/AppPaciente.java
package br.unifor.facilitador.boundary;

import br.unifor.facilitador.dominio.*;
import br.unifor.facilitador.porta.*;
import java.time.LocalDateTime;
import java.util.UUID;

/**
 * Fronteira do sistema (nao e entidade de dominio).
 * Repare que ela depende SO de upload: nao tem acesso a download nem a triagem.
 */
public class AppPaciente {

    private final AutenticacaoService auth;
    private final AgendaService agenda;
    private final DocumentoUploadPort upload;
    private Sessao sessao;

    public AppPaciente(AutenticacaoService auth, AgendaService agenda, DocumentoUploadPort upload) {
        this.auth = auth;
        this.agenda = agenda;
        this.upload = upload;
    }

    public void entrar(String email, String senha) {
        this.sessao = auth.login(email, senha);
        System.out.println("[app-paciente] login ok, token valido ate " + sessao.getExpiraEm());
    }

    public Consulta agendarConsulta(Paciente p, Medico m, LocalDateTime dt, Modalidade mod) {
        exigirSessao();
        return agenda.agendar(p, m, dt, mod);
    }

    public UUID enviarExamePrevio(Consulta c, String uri, UUID autor) {
        exigirSessao();
        Documento d = new Documento(TipoDocumento.EXAME_PREVIO, uri, autor);
        c.anexar(d);
        return upload.enviar(d);
    }

    private void exigirSessao() {
        if (sessao == null || !sessao.estaValida()) {
            throw new SecurityException("Faca login antes.");
        }
    }
}
